### PR TITLE
Do not discard context from substeps

### LINKED ISSUE #487: Context from substeps is lost
the `suite.maybeSubSteps` function loops over all substeps and invoked them one by one:

https://github.com/cucumber/godog/blob/5e176da4338ad55ca074f7932b1b5dd38819890f/suite.go#L349-L356

On line 352 new `ctx` and `err` variables are created in the scope of the `if` statement that shadow the `ctx` from the argument list. As soon as the if scope completes `ctx` becomes the original context that was passed in to the function again. The effect of this is that the context returned by the subset is discarded.

This can easily be tested by modifying the standard example to add a multistep:

```go
func InitializeScenario(sc *godog.ScenarioContext) {
	sc.StepContext().Before(transformStep)
	sc.Step(`^I eat (\d+)$`, iEat)
	sc.Step(`^there should be (\d+) remaining$`, thereShouldBeRemaining)
	sc.Step(`^I eat (\d+) godogs out of (\d+)`, func(ctx context.Context, a, b int) (context.Context, godog.Steps) {
		return ctx, godog.Steps{
			fmt.Sprintf("there are %d godogs", b),
			fmt.Sprintf("I eat %d", a),
		}
	})
}
```

with a matching scenario:

```gherkin
    Scenario: Basic multistep
        Given I eat 8 godogs out of 14
        Then there should be 6 remaining
```

### PR DESCRIPTION
**Pull request by wichert**
The context from a substep was accidentally discarded because the `ctx` variable was shadowed.

This fixes #487

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://cucumber.io/conduct/)
- [X] I've changed the behaviour of the code
  - [x] I have added/updated tests to cover my changes.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [x] Users should know about my change
  - [x] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../CHANGELOG.md), linking to this pull request.

----

*This text was originally generated from a [template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates), then edited by hand. [You can modify the template here.](https://github.com/cucumber/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

### CODE REVIEW

**Review by vearutop — APPROVED** [review]
👍

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ This document is formatted according to the principles of [Keep A CHANGELOG](htt
 - README example is updated with `context.Context` and `go test` usage. ([477](https://github.com/cucumber/godog/pull/477) - [vearutop](https://github.com/vearutop)) 
 
 ### Fixed
+- Fixed a bug which would ignore the context returned from a substep.([488](https://github.com/cucumber/godog/pull/480) - [wichert](https://github.com/wichert))
```
Comment (vearutop) [inline CHANGELOG.md:15]:
```suggestion
- Fixed a bug which would ignore the context returned from a substep.([488](https://github.com/cucumber/godog/pull/488) - [wichert](https://github.com/wichert))
```

Comment (wichert) [reply CHANGELOG.md]: Thanks, fixed!

**Comment by wichert** [comment]
I'm afraid I don't understand why the tests are failing. The errors look unrelated to my changes?

**Comment by dumpsterfireproject** [comment]
I had a similar experience with my commit.  In run_test.go, Test_AllFeaturesRun and Test_AllFeaturesRunAsSubtests run all the .features in the features directory with the progress formatter. So when new .features are added to that directory, it changes the expected output of the progress formatter.  The number of dots output by the progress formatter as well as the scenario and step counts output will all increase with the new .feature added to that directory. I updated those tests to reflect the additional steps & scenarios on my PR. But we should probably log an issue to change the Options.Paths used in those two tests to only use .features in that directory that the test expects instead of just running them all, so the test doesn't break every time a new .feature is added.

**Comment by vearutop** [comment]
Yes, that's true. In order to fix tests you need to update [expected report](https://github.com/cucumber/godog/blob/v0.12.5/run_test.go#L426-L468) (which is now having a few more steps covered).

```
    run_test.go:483: 
        	Error Trace:	run_test.go:483
        	Error:      	Not equal: 
        	            	expected: "...................................................................... 70\n...................................................................... 140\n...................................................................... 210\n...................................................................... 280\n.................................................                      329\n\n\n86 scenarios (86 passed)\n329 steps (329 passed)\n0s\n"
        	            	actual  : "...................................................................... 70\n...................................................................... 140\n...................................................................... 210\n...................................................................... 280\n.......................................................                335\n\n\n88 scenarios (88 passed)\n335 steps (335 passed)\n0s\n"
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -4,7 +4,7 @@
        	            	 ...................................................................... 280
        	            	-.................................................                      329
        	            	+.......................................................                335
        	            	-86 scenarios (86 passed)
        	            	-329 steps (329 passed)
        	            	+88 scenarios (88 passed)
        	            	+335 steps (335 passed)
        	            	 0s

```

**Comment by wichert** [comment]
Thanks, that solved the test errors!

FWIW when using go 1.18 the tests are failing because the source filename is no longer available, resulting in many instances of this:

```
 Scenario: load features within path    # features/load.feature:6
-Given a feature path "features"      # suite_context_test.go:0 -> *godogFeaturesScenario
-When I parse features                # suite_context_test.go:0 -> *godogFeaturesScenario
-Then I should have 13 feature files: # suite_context_test.go:0 -> *godogFeaturesScenario
+Given a feature path "features"      # <autogenerated>:1 -> *godogFeaturesScenario
+When I parse features                # <autogenerated>:1 -> *godogFeaturesScenario
+Then I should have 13 feature files: # <autogenerated>:1 -> *godogFeaturesScenario
```

It looks like the name of the source files is lost, and replaced with `<autogenerated>`

**Comment by vearutop** [comment]
Yes, go1.18 has a technically breaking change wrt closure introspection (https://github.com/cucumber/godog/pull/466, https://github.com/golang/go/issues/51774), hopefully it will be fixed in 1.19 or 1.20.

On our side I think we should also upgrade function resolver to not show `autogenerated` as it is useless.

**Comment by aslakhellesoy** [comment]
Hi @wichert,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Pull request merged**